### PR TITLE
[SYCLLowerIR] Add support for FPGA latency control extension

### PR DESCRIPTION
**Pull request by asudarsa**
This PR adds support to lower compile time properties and generate llvm pointer annotations ready to be consumed by SPIR-V gen.
The compile-time properties lowered are latency-anchor-id and latency-constraint.

Thanks

### CODE REVIEW

**Comment by asudarsa** [comment]
Hi @steffenlarsen 

Can you please take a look?

Thanks

**Review by steffenlarsen — APPROVED** [review]
Looks good to me!

Only nit I have is that I would prefer the test to be a little more focused, that is it has a lot of unnecessary information which could be removed without affecting the test itself.

File: llvm/test/SYCLLowerIR/CompileTimePropertiesPass/sycl-latency-control.ll [hunk 1]
```diff
@@ -7,60 +7,57 @@ target triple = "spir64-unknown-unknown"
 
 $_ZTSZ4fooEUlvE_ = comdat any
 
-@.str = private unnamed_addr addrspace(1) constant [16 x i8] c"sycl-properties\00", section "llvm.metadata"
```
Comment (asudarsa) [inline llvm/test/SYCLLowerIR/CompileTimePropertiesPass/sycl-latency-control.ll]: I removed addrspace(1) to make the test simpler.

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 2]
```diff
@@ -41,6 +41,7 @@ constexpr uint32_t SPIRV_PIPELINE_ENABLE_DECOR = 5919;
 enum class DecorValueTy {
   uint32,
   boolean,
+  string,
```
Comment (asudarsa) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:44]: Added a string DecorValueTy to enabling the store of a list of literals as a comma-separated string.

**Comment by asudarsa** [comment]
> Looks good to me!
> 
> Only nit I have is that I would prefer the test to be a little more focused, that is it has a lot of unnecessary information which could be removed without affecting the test itself.

I modified the test to have opaque pointers and also cleaned it up a bit.

**Comment by asudarsa** [comment]
@intel/dpcpp-tools-reviewers 

Can you please take a look? 

Thanks

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 3]
```diff
@@ -491,9 +519,14 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
...
-  if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
-    if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+  if (IntrAnnotStringArg->getType()->isOpaquePointerTy()) {
+    if (auto *C = dyn_cast<Constant>(IntrAnnotStringArg))
       AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  } else {
+    if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
+      if (auto *C = dyn_cast<Constant>(GEP->getOperand(0)))
+        AnnotStrArgGV = dyn_cast<GlobalVariable>(C);
+  }
```
Comment (steffenlarsen) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:529]: Nit;
```suggestion
  if (IntrAnnotStringArg->getType()->isOpaquePointerTy())
    AnnotStrArgGV = dyn_cast<GlobalVariable>(IntrAnnotStringArg);
  else if (auto *GEP = dyn_cast<GEPOperator>(IntrAnnotStringArg))
    AnnotStrArgGV = dyn_cast<GlobalVariable>(GEP->getOperand(0));
```

**Review by steffenlarsen — APPROVED** [review]
Looks good to me! just a few minor simplification nits

**Comment by asudarsa** [comment]
Test failures are not related to this change and are found to be failing in other locations as well. Thanks

**Comment by asudarsa** [comment]
Hi @sarnex 

Can you please help to take a look at this change and help to merge?

Thanks

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 4]
```diff
@@ -77,6 +78,26 @@ MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
...
+/// @returns a pointer to the metadata node created for the required decoration
+/// and its value.
+MDNode *buildSpirvDecorMetadata(LLVMContext &Ctx, uint32_t OpCode,
+                                StringRef Value) {
+  auto *Ty = Type::getInt32Ty(Ctx);
+  SmallVector<Metadata *, 2> MD;
+  MD.push_back(ConstantAsMetadata::get(
+      Constant::getIntegerValue(Ty, APInt(32, OpCode))));
+  MD.push_back(
+      ConstantAsMetadata::get(ConstantDataArray::getString(Ctx, Value, true)));
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:97]: nit: can we add a comment on what true represents, like
```
ConstantDataArray::getString(Ctx, Value, /*whatever=*/true)
```

**Review by sarnex — APPROVED** [review]

**Pull request merged**